ledger/complete/trigger.go

```diff
@@ -0,0 +1,142 @@
+package complete
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/ledger/complete/common"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+// CheckpointRunner abstraction for the runner, so that easy to test Trigger
+type CheckpointRunner interface {
+	RunCheckpoint(ctx context.Context, checkpointTries []*trie.MTrie, checkpointNum int) error
+}
+
+// Trigger takes checkpointing related configs and decides when to trigger checkpointing,
+// if triggered, it will use the checkpointRunner to create checkpoint.
+// It ensures only one checkpoint is being created at any time.
+type Trigger struct {
+	logger                  zerolog.Logger
+	activeSegmentNum        int                     // the num of the segment that the previous trie update was written to, is used to detect whether trie update is written to a new segment.
+	checkpointedSegumentNum *atomic.Int32           // the highest segment num that has been checkpointed
+	checkpointDistance      uint                    // the total number of segment (write-ahead log) to trigger checkpointing
+	isRunning               *atomic.Bool            // an atomic.Bool for whether checkpointing is currently running.
```

Comment:
Just out of curiosity, is there a benefit of using atomic.Bool instead `semaphore.NewWeighted(1)` ?

Reply:
No big difference. However,I think semaphore is mostly useful in a scenario where Weight is above 1. 

In our specific scenario here, Weight is 1, as we only want at most a single thread to run checkpointing. So it makes more sense simply using an atomic.Bool. A semaphore is a little bit overkill. 

With the isRunning state as an atomic.Bool value, if we only allow one thread to change this value, and only the thread changed isRunning to true to run the checkpointing, and we set isRunning back to false after checkpointing is done, Then we can guarantee checkpointing is run by at most one thread at any time.